x-pack/metricbeat/module/azure/client_batch.go

```diff
@@ -0,0 +1,469 @@
+}
+
+// GetVMForMetadata func will retrieve the VM details in order to fill in the cloud metadata
+// and also update the client resources
+func (client *BatchClient) GetVMForMetadata(resource *Resource, referencePoint KeyValuePoint) VmResource {
+	var (
+		vm           VmResource
+		resourceName = resource.Name
+		resourceId   = resource.Id
+	)
+
+	// Search the dimensions for the "VMName" dimension. This dimension is present for VM Scale Sets.
+	if dimensionValue, ok := getDimension("VMName", referencePoint.Dimensions); ok {
+		instanceId := getInstanceId(dimensionValue)
+		if instanceId != "" {
+			resourceId += fmt.Sprintf("/virtualMachines/%s", instanceId)
+			resourceName = dimensionValue
+		}
+	}
+
+	// if vm has been already added to the resource then it should be returned
+	if existingVM, ok := getVM(resourceName, resource.Vms); ok {
+		return existingVM
+	}
+
+	// an additional call is necessary in order to retrieve the vm specific details
+	expandedResource, err := client.AzureMonitorService.GetResourceDefinitionById(resourceId)
+	if err != nil {
+		client.Log.Error(err, "could not retrieve the resource details by resource ID %s", resourceId)
+		return VmResource{}
+	}
+
+	vm.Name = *expandedResource.Name
+
+	if expandedResource.Properties != nil {
+		if properties, ok := expandedResource.Properties.(map[string]interface{}); ok {
+			if hardware, ok := properties["hardwareProfile"]; ok {
+				if vmSz, ok := hardware.(map[string]interface{})["vmSize"]; ok {
+					vm.Size = vmSz.(string)
+				}
+				if vmID, ok := properties["vmId"]; ok {
+					vm.Id = vmID.(string)
+				}
+			}
+		}
+	}
+
+	if len(vm.Size) == 0 && expandedResource.SKU != nil && expandedResource.SKU.Name != nil {
+		vm.Size = *expandedResource.SKU.Name
+	}
+
+	// the client resource and selected resources are being updated in order to avoid additional calls
+	client.AddVmToResource(resource.Id, vm)
+
+	resource.Vms = append(resource.Vms, vm)
+
+	return vm
+}
+
+// GetResourceForMetaData will retrieve resource details for the selected metric configuration
+func (client *BatchClient) GetResourceForMetaData(grouped Metric) Resource {
+	for _, res := range client.Resources {
+		if res.Id == grouped.ResourceId {
+			return res
+		}
+	}
+	return Resource{}
+}
+
+func (client *BatchClient) LookupResource(resourceId string) Resource {
+	for _, res := range client.Resources {
+		if res.Id == resourceId {
+			return res
+		}
+	}
+	return Resource{}
+}
+
+// AddVmToResource will add the vm details to the resource
+func (client *BatchClient) AddVmToResource(resourceId string, vm VmResource) {
+	if len(vm.Id) > 0 && len(vm.Name) > 0 {
+		for i, res := range client.Resources {
+			if res.Id == resourceId {
+				client.Resources[i].Vms = append(client.Resources[i].Vms, vm)
+			}
+		}
+	}
+}
+
+// mapToEvents maps the metric values to events and reports them to Elasticsearch.
+func (client *BatchClient) MapToEvents(metrics []Metric, reporter mb.ReporterV2) error {
```

Comment:
Besides `InitResources()`, `GroupAndStoreMetrics()`, and `GetMetricsInBatch()`, there are other functions likes `MapToEvents()`, `AddVmToResource()` that seems to contain the same code as the non-batch counterpart.

Reply:
I updated the code by using a baseClient which implements all the common methods